### PR TITLE
Introduce `tryCapture(_:)`.

### PR DESCRIPTION
**Pull request by rxwei**
`tryCapture(_:)` works like `capture(_:)` but accepts an optional or throwing capture transform instead. When the transform fails (returns `nil` or throws), it will cause matching to fail. This works well with failable conversion initializers such as `Int.init?(_:)`, `Unicode.Scalar.init?(_:)`, etc.

----

Example:

With `tryCapture`, the following example avoids unnecessary levels of optionals and double optionals.

```swift
line.match {
  OneOrMore(.hexDigit).tryCapture(Unicode.Scalar.init(hex:))
  Optionally {
    ".."
    OneOrMore(.hexDigit).tryCapture(Unicode.Scalar.init(hex:))
  }
  OneOrMore(.whitespace)
  ";"
  OneOrMore(.whitespace)
  OneOrMore(.word).tryCapture(Unicode.GraphemeBreakProperty.init)
  Repeat(.any)
} // Regex<(Substring, Unicode.Scalar, Unicode.Scalar?, Unicode.GraphemeBreakProperty)>
```

### CODE REVIEW

**Comment by rxwei** [comment]
Note: This PR depends on #120 and #119. Please review 36c1a03 for changes specific to `tryCapture(_:)`.

**Comment by milseman** [comment]
Should we have a throw abort? That's the pattern we're planning to follow elsewhere and Perl/PCRE's more general-form callouts have that capability.

File: Sources/Exercises/Participants/RegexParticipant.swift [hunk 1]
```diff
@@ -63,28 +65,34 @@ private func graphemeBreakPropertyData<RP: RegexProtocol>(
...
-private func graphemeBreakPropertyDataLiteral(
-  forLine line: String
-) -> GraphemeBreakEntry? {
-  return graphemeBreakPropertyData(
-    forLine: line,
-    using: r(#"([0-9A-F]+)(?:\.\.([0-9A-F]+))?\s+;\s+(\w+).*"#,
-             matching: Tuple4<Substring, Substring, Substring?, Substring>.self))
+  }.map {
+    let (_, lower, upper, property) = $0.match.tuple
+    print(property)
```
Comment (milseman) [inline Sources/Exercises/Participants/RegexParticipant.swift:95]:
```suggestion
```

File: Sources/_MatchingEngine/Regex/AST/AST.swift [hunk 2]
```diff
@@ -207,14 +210,61 @@ extension AST {
...
+  }
+
+  public init(
+    resultType: Any.Type,
+    _ closure: @escaping (Substring) throws -> Any
+  ) {
+    self.init(resultType: resultType, closure: .throwing(closure))
+  }
+
+  public func callAsFunction(_ input: Substring) -> Any? {
```
Comment (milseman) [inline Sources/_MatchingEngine/Regex/AST/AST.swift:247]: Should this be `throws` then?

File: Tests/RegexTests/RegexDSLTests.swift [hunk 3]
```diff
@@ -233,3 +316,13 @@ class RegexDSLTests: XCTestCase {
     }
   }
 }
+
+extension Unicode.Scalar {
+  // Convert a hexadecimal string to a scalar
+  public init?<S: StringProtocol>(hex: S) {
```
Comment (milseman) [inline Tests/RegexTests/RegexDSLTests.swift:322]:
```suggestion
 init?<S: StringProtocol>(hex: S) {
```

**Review by milseman — APPROVED** [review]
Overall LGTM!

File: Sources/_MatchingEngine/Regex/AST/AST.swift (hunk 2 above)
Comment (rxwei) [reply Sources/_MatchingEngine/Regex/AST/AST.swift:247]: No, this returns an optional for the engine to determine whether to fail matching. When the closure throws, this returns nil.

Comment (rxwei) [reply Sources/_MatchingEngine/Regex/AST/AST.swift:247]: > Should we have a throw abort?
Just saw this comment. In that case I think we can have if.

Comment (milseman) [reply Sources/_MatchingEngine/Regex/AST/AST.swift:247]: SGTM

**Comment by milseman** [comment]
If this passes your local testing, I'm ok merging this.

@shahmishal any idea why Linux CI can't get us a machine? We run in around 1min total time, so we're not going to clog up any pipelines.

**Comment by rxwei** [comment]
I haven't triggered tests here, as #119 needed to be merged first.

**Comment by rxwei** [comment]
@swift-ci please test Linux

**Pull request merged**